### PR TITLE
Rewrite addon-resizer nanny

### PR DESCRIPTION
**Pull request by x13n**
There will be a single estimator class, providing two ranges: acceptable
and recommended range. As long as current pod requirements and limits fall into
acceptable range, nothing happens. Once either limits or requirements
fall out of acceptable range, they are both updated to lower (when
upscaling) or higher (when downscaling) end of recommended range. This
approach prevents flapping, which took place in previous implementation,
when cluster size oscillated around certain values.

More details: https://docs.google.com/a/google.com/document/d/1T0A7GHwNU_w6gpq_eCN166aRth6usbRvxgJJqSZeESU/edit?usp=sharing (shared with kubernetes-sig-instrumentation and kubernetes-sig-autoscaling)

Also, fix the code to actually use poll-period flag.

(PR moved from https://github.com/kubernetes/contrib/pull/2623)

### CODE REVIEW

File: addon-resizer/Makefile [hunk 1]
```diff
@@ -25,7 +25,7 @@
 OUT_DIR = build
 PACKAGE = k8s.io/contrib/addon-resizer
 PREFIX = gcr.io/google_containers
-TAG = 1.7
+TAG = 1.8
```
Comment (mwielgus) [inline addon-resizer/Makefile]: it is a significant rewrite - lets bump it to 2.0

File: addon-resizer/nanny/estimator.go [hunk 2]
```diff
@@ -18,50 +18,59 @@ package nanny
 
 import (
 	"fmt"
+	"math"
 
+	log "github.com/golang/glog"
```
Comment (mwielgus) [inline addon-resizer/nanny/estimator.go]: imports:

core-go
<blank line>
kubernetes
<blank line>
3rd-party

File: addon-resizer/Makefile (hunk 1 above)
Comment (x13n) [inline addon-resizer/Makefile]: Ok.

File: addon-resizer/nanny/estimator.go (hunk 2 above)
Comment (x13n) [inline addon-resizer/nanny/estimator.go]: And your point is...? :-) Note that nanny_lib was already using glog.

Comment (mwielgus) [inline addon-resizer/nanny/estimator.go]: wrong import order

Comment (x13n) [inline addon-resizer/nanny/estimator.go]: Oh, you meant the order. Fixed.

File: addon-resizer/nanny/estimator.go [hunk 3]
```diff
@@ -34,34 +32,45 @@ type Resource struct {
 	Name               api.ResourceName
 }
 
-// LinearEstimator estimates the amount of resources as r = base + extra*nodes.
-type LinearEstimator struct {
-	Resources []Resource
+// ResourceListPair is a pair of ResourceLists, denoting a range.
+type ResourceListPair struct {
+	lower, upper api.ResourceList
```
Comment (mwielgus) [inline addon-resizer/nanny/estimator.go]: what exactly is lower/upper? is it min/max? please document.

File: addon-resizer/nanny/estimator.go [hunk 4]
```diff
@@ -34,34 +32,45 @@ type Resource struct {
...
 }
 
-func (e LinearEstimator) scaleWithNodes(numNodes uint64) *api.ResourceRequirements {
-	return calculateResources(numNodes, e.Resources)
+// EstimatorResult is the result of the resource Estimation, used by Estimator struct.
+type EstimatorResult struct {
+	RecommendedRange, AcceptableRange ResourceListPair
```
Comment (mwielgus) [inline addon-resizer/nanny/estimator.go]: All public fields must have documentation.
One field per line.

File: addon-resizer/nanny/estimator.go [hunk 5]
```diff
@@ -34,34 +32,45 @@ type Resource struct {
...
 
-// ExponentialEstimator estimates the amount of resources in the way that
-// prevents from frequent updates but may end up with larger resource usage
-// than actually needed (though no more than ScaleFactor).
-type ExponentialEstimator struct {
-	Resources   []Resource
-	ScaleFactor float64
+// Estimator is a struct used for estimating accepted and recommended resource requirements.
+type Estimator struct {
+	Resources            []Resource
```
Comment (mwielgus) [inline addon-resizer/nanny/estimator.go]: Doc.

File: addon-resizer/nanny/main/pod_nanny.go [hunk 6]
```diff
@@ -34,37 +35,58 @@ const noValue = "MISSING"
...
 	// Flags to identify the container to nanny.
 	podNamespace  = flag.String("namespace", os.Getenv("MY_POD_NAMESPACE"), "The namespace of the ward. This defaults to the nanny pod's own namespace.")
 	deployment    = flag.String("deployment", "", "The name of the deployment being monitored. This is required.")
 	podName       = flag.String("pod", os.Getenv("MY_POD_NAME"), "The name of the pod to watch. This defaults to the nanny's own pod.")
 	containerName = flag.String("container", "pod-nanny", "The name of the container to watch. This defaults to the nanny itself.")
 	// Flags to control runtime behavior.
-	pollPeriod = time.Millisecond * time.Duration(*flag.Int("poll-period", 10000, "The time, in milliseconds, to poll the dependent container."))
-	estimator  = flag.String("estimator", "linear", "The estimator to use. Currently supported: linear, exponential")
+	pollPeriodMillis = flag.Int("poll-period", 10000, "The time, in milliseconds, to poll the dependent container.")
+	estimator        = flag.String("estimator", "linear", "[DEPRECATED] No-op. Can be set to: linear, exponential")
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: Remove.

File: addon-resizer/nanny/main/pod_nanny.go [hunk 7]
```diff
@@ -34,37 +35,58 @@ const noValue = "MISSING"
...
-	baseStorage    = flag.String("storage", noValue, "The base storage resource requirement.")
-	storagePerNode = flag.String("extra-storage", "0Gi", "The amount of storage to add per node.")
-	threshold      = flag.Int("threshold", 0, "A number between 0-100. The dependent's resources are rewritten when they deviate from expected by more than threshold.")
+	baseCPU              = flag.String("cpu", noValue, "The base CPU resource requirement.")
+	cpuPerNode           = flag.String("extra-cpu", "0", "The amount of CPU to add per node.")
+	baseMemory           = flag.String("memory", noValue, "The base memory resource requirement.")
+	memoryPerNode        = flag.String("extra-memory", "0Mi", "The amount of memory to add per node.")
+	baseStorage          = flag.String("storage", noValue, "The base storage resource requirement.")
+	storagePerNode       = flag.String("extra-storage", "0Gi", "The amount of storage to add per node.")
+	threshold            = flag.Int("threshold", 0, "[DEPRECATED] Please use recommendation-offset and acceptance-offset istead.\nA number between 0-100. The dependent's resources are rewritten when they deviate from expected by more than threshold.")
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: Remove.

File: addon-resizer/nanny/main/pod_nanny.go [hunk 8]
```diff
@@ -34,37 +35,58 @@ const noValue = "MISSING"
...
 	podName       = flag.String("pod", os.Getenv("MY_POD_NAME"), "The name of the pod to watch. This defaults to the nanny's own pod.")
 	containerName = flag.String("container", "pod-nanny", "The name of the container to watch. This defaults to the nanny itself.")
 	// Flags to control runtime behavior.
-	pollPeriod = time.Millisecond * time.Duration(*flag.Int("poll-period", 10000, "The time, in milliseconds, to poll the dependent container."))
-	estimator  = flag.String("estimator", "linear", "The estimator to use. Currently supported: linear, exponential")
+	pollPeriodMillis = flag.Int("poll-period", 10000, "The time, in milliseconds, to poll the dependent container.")
+	estimator        = flag.String("estimator", "linear", "[DEPRECATED] No-op. Can be set to: linear, exponential")
 )
 
+func checkPercentageBounds(flagName string, flagValue int) {
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: checkPercentageFlagBounds

File: addon-resizer/nanny/main/pod_nanny.go [hunk 9]
```diff
@@ -34,37 +35,58 @@ const noValue = "MISSING"
...
+		*acceptanceOffset = *threshold
 	}
 
+	checkPercentageBounds("recommendation-offset", *recommendationOffset)
+	checkPercentageBounds("acceptance-offset", *acceptanceOffset)
+
+	pollPeriod := time.Millisecond * time.Duration(*pollPeriodMillis)
+	log.Infof("Poll period: %+v", pollPeriod)
 	log.Infof("Watching namespace: %s, pod: %s, container: %s.", *podNamespace, *podName, *containerName)
 	log.Infof("cpu: %s, extra_cpu: %s, memory: %s, extra_memory: %s, storage: %s, extra_storage: %s", *baseCPU, *cpuPerNode, *baseMemory, *memoryPerNode, *baseStorage, *storagePerNode)
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: What about printing bounds?

File: addon-resizer/nanny/main/pod_nanny.go [hunk 10]
```diff
@@ -109,20 +131,21 @@ func main() {
...
-	if *estimator == "linear" {
-		est = nanny.LinearEstimator{
-			Resources: resources,
-		}
-	} else if *estimator == "exponential" {
-		est = nanny.ExponentialEstimator{
-			Resources:   resources,
-			ScaleFactor: 1.5,
-		}
+	if *estimator == "linear" || *estimator == "exponential" {
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: AFAIR this was about to be a noop.

Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go]: Remove.

File: addon-resizer/nanny/nanny_lib.go [hunk 11]
```diff
@@ -24,39 +24,47 @@ import (
...
-		checkResource(threshold, limits, expLimits, api.ResourceStorage) ||
-		checkResource(threshold, reqs, expReqs, api.ResourceCPU) ||
-		checkResource(threshold, reqs, expReqs, api.ResourceMemory) ||
-		checkResource(threshold, reqs, expReqs, api.ResourceStorage)
+// Returns null when no resources should be overriden.
+// Otherwise, returns ResourceList that should be used.
+func shouldOverwriteResources(estimatorResult *EstimatorResult, limits, reqs api.ResourceList) *api.ResourceRequirements {
+	for _, list := range []api.ResourceList{limits, reqs} {
+		for _, resourceType := range []api.ResourceName{api.ResourceCPU, api.ResourceMemory, api.ResourceStorage} {
+			p := checkResource(estimatorResult, list, resourceType)
```
Comment (mwielgus) [inline addon-resizer/nanny/nanny_lib.go]: what is p. Please use full names.

**Comment by mwielgus** [comment]
Please explain how the PR was tested.

**Comment by x13n** [comment]
Apart from unit tests, I ran a manual integration test, setting up my own Kubernetes cluster on GCE and resizing it in a way described in the doc (example scenario) linked in the first post & verifying changes occur as expected.

**Comment by mwielgus** [comment]
Doesn't compile.

**Comment by x13n** [comment]
Does compile.

**Comment by mwielgus** [comment]
Fixes https://github.com/kubernetes/kubernetes/issues/30446.

**Comment by mwielgus** [comment]
LGTM

**Pull request merged**